Group iteration on simplifying principle wording (#14)
open-gitops/documents: PRINCIPLES.md

## Principles

> 1. **Desired System State Is Declarative**

Reviewer: ```suggestion
1. **Desired System State is Declarative**
```
Typographic Suggestion:
`Is` -> `is`
to align with
https://user-images.githubusercontent.com/407675/126000898-eb81abd1-a947-40e4-b313-107fd45b7ad7.png
Thanks @scottrigby

Author: See https://github.com/open-gitops/documents/pull/14#discussion_r676641157

    A system managed by GitOps must have its _Desired State_ expressed declaratively as data in a format writable and readable by both humans and machines.

> 2. **Declarations Are Versioned and Immutable**

Reviewer: ```suggestion
2. **Declarations are Versioned and Immutable**
```
Typographic Suggestion:
`Are` -> `are`
to align with
https://user-images.githubusercontent.com/407675/126000898-eb81abd1-a947-40e4-b313-107fd45b7ad7.png
Thanks @scottrigby

Author: @lloydchang Yeah I wasn't sure about which title case style we want to adopt. I would recommend not getting too caught up in this question, and that we follow some existing style guide. For now I just chose AP style [AP style Title Case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case) so I didn't have to think about it.

If interested though, this is a fun tool to see how the different standards look https://titlecaseconverter.com/

> 2. **State declarations are stored as immutable versions**

Author: ```suggestion
2. **Declarations are Version Controlled**
```

Reviewer: ```suggestion
2. **Declarations are Versioned**
```

Reviewer: 🤏

Author: @lloydchang Thanks!

Re https://github.com/open-gitops/documents/pull/14#pullrequestreview-708672508, I do personally agree the shorter and more memorable/catchy we can make these principle headers the better 🤏 

I wonder though if "versioned" here is enough, whereas "version controlled" seems to me to check all the boxes of this principle. But do you not feel these have different meanings?

> Desired State is stored in a way that supports _versioning_, _immutability of versions_, and _retains a complete version history_.

Author: @jlbutler 

```suggestion
2. **Declarations are versioned and Immutable**
```

Reviewer 2: ```suggestion
2. **Declarations are audit-proof versioned**
```

Author: @jlbutler Updated the PR. 

@heubeck thanks for this suggestion! I haven't added this to the branch yet because I think we'd need more discussion on that first. but I added you as co-author to this commit as well since you're collaborating with us here on GitHub. I'd like us to get this PR merged soon, but we can continue iterating. Are you free to join the next principles meeting on Wednesday?

Reviewer 2: @scottrigby Thank you, I'll try to participate.

Reviewer 2: Regarding my `audit-proof` suggestion: Technically it means the same as the phrases with `immutable` and `versions`, but the statement is much stronger.
Industries with deployment revision logging ("audit") demands may appreciate a less technical terminology.

Author: @heubeck this is really interesting. Given how close this PR is to being merged for `v0.2.0`, and how important it is to simplify the headers ASAP, let's discuss this idea for a next iteration 🙏 

Would you be willing to move this idea to a [discussion in the OpenGitOps project repo](https://github.com/open-gitops/project/discussions)? That way there is a place for async discussion on this before and after the next meeting (tomorrow). The discussion could recommend this change from the current PR, and see how that sits with the principles committee and community?
```diff
- 2. **Declarations Are Versioned and Immutable**
+ 2. **Declarations Are Versioned and Audit-Proof**
```

Reviewer 2: Happy to discuss: https://github.com/open-gitops/project/discussions/15

    _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.

> 3. **State Reconciliation Is Continuous**

Reviewer: ```suggestion
3. **State Reconciliation is Continuous**
```
Typographic Suggestion:
`Is` -> `is`
to align with
https://user-images.githubusercontent.com/407675/126000898-eb81abd1-a947-40e4-b313-107fd45b7ad7.png
Thanks @scottrigby

Author: See https://github.com/open-gitops/documents/pull/14#discussion_r676641157

> 3. **State reconciliation is automated and continuous**

Author: Just changing the order to match the sentence below.
```suggestion
3. **State is Continuously and Automatically Reconciled**
```

    Software agents continuously, and automatically, compare a system's state to its _Desired State_.

>     If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

Author: Good point. We could even just remove repetition altogether and say:

```suggestion
    If these differ for any reason, automated actions to reconcile the system are initiated.
```

This made me realize we may want to be specific about what is being reconciled when there's a difference between the two. It's the system itself that gets reconciled, not the desired state 🙃

>     If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

Reviewer: No, @scottrigby, because desired state can be reconciled; please see https://youtu.be/fZEFWh4xBEw?t=1135
> "Convergence doesn't just mean left to right.  It might in fact mean right to left.  It might mean there is some kind of automation that's happening in the runtime system that is doing legitimate things and I'll give you a concrete example of later in the talk, that is doing legitimate things that then need to be reflected back into the git repository.  Because not only is the git repository, not only is git the interface for operation, it now becomes the source of truth; it is the desired state, so if we're constantly bringing these things into alignment; if something happens in the runtime environment that should be part of the desired state, we need to ensure that is captured as well."

>     If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

Author: @lloydchang I have been a bit hung up on terminology of convergence vs reconciliation. However after talking with @cdavisafc about this more I agree with you. There is more to be said about the details (like write back agents following rules specified in the intentional declarations, not mixing intentional and automated write back info etc), but that can be addressed in a glossary item, or in best practices or another supporting doc 👍  Updated and added your as co-author

>     If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

Reviewer: @scottrigby @cdavisafc Thank you both!

> 4. **Intentional change happens only through changing state declarations**

Author: ```suggestion
4. **System Changes are Only Through Version Control**
```

Reviewer: ```suggestion
4. **Changes are Only Through Version Control**
```

Reviewer: 🤏

Author: Hm, my gut response to "Changes are Versioned" for principle 4 is this crosses the line from brevity to loosing the intended meaning 🤔 🧠 

I think the main point of 4 is that – in principle – for a system to be considered GitOps-managed it must only use new versions of desired state declarations for all intentional changes to the system, not as one of many methods to change the system. Right?

There are exceptions noted - such as the example of "break glass" for manual intervention during incident management etc – during which time, this GitOps principle is considered to be temporarily suspended.

WDYT?

Reviewer 3: "Updates to desired state are versioned"

Author: @moshloop suggested:

```suggestion
4. **Intentional Changes Only Through Versioning**
```

> 4. **Intentional Changes Only Through Versioning**

Reviewer 3: What about `4. Actualizations to Declarations are versioned`

Reviewer 3: Note: Not sure if these are interchangeable in this context `Actualizations to Declarations...` vs `Updates to Declarations...` vs `Changes to Declarations...`

Author: 🤔

> 4. **Intentional Changes Only Through Versioning**

Author: To address grammar:

```suggestion
4. **Intentional Changes are Only Through Versioning**
```

Author: 🤔

Reviewer: ```suggestion
4. **GitOps intends these Operating principles, so.**
```
Recursive acronym 🙂

WDYT, @scottrigby et al.?
